src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -126,6 +138,93 @@ class AzureService(crlService: CrlService, directoryDAO: DirectoryDAO, azureMana
       createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
     } yield (createdPet, true)
 
+  def getOrCreateActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      billingProfileId: BillingProfileId,
+      samUser: SamUser,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] =
+    for {
+      hasPermission <- policyEvaluatorService.hasPermission(resource, resourceAction, samUser.id, samRequestContext)
+      _ <- IO.raiseWhen(!hasPermission)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, s"User ${samUser.id} does not have $resourceAction on $resource"))
+      )
+      id = ActionManagedIdentityId(resource, resourceAction, billingProfileId)
+      existingAmiOpt <- directoryDAO.loadActionManagedIdentity(id, samRequestContext)
+      ami <- existingAmiOpt match {
+        // pet exists in Sam DB - return it
+        case Some(p) => IO.pure((p, false))
+        // pet does not exist in Sam DB - create it
+        case None => createActionManagedIdentity(id, samRequestContext)
+      }
+    } yield ami
+
+  def getActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      samUser: SamUser,
+      samRequestContext: SamRequestContext
+  ): IO[Option[ActionManagedIdentity]] =
+    for {
+      hasPermission <- policyEvaluatorService.hasPermission(resource, resourceAction, samUser.id, samRequestContext)
+      _ <- IO.raiseWhen(!hasPermission)(
+        new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Forbidden, s"User ${samUser.id} does not have $resourceAction on $resource"))
+      )
+      ami <- directoryDAO.loadActionManagedIdentity(resource, resourceAction, samRequestContext)
+    } yield ami
+
+  private def createActionManagedIdentity(
+      id: ActionManagedIdentityId,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] =
+    for {
+      mrgOpt <- azureManagedResourceGroupDAO.getManagedResourceGroupByBillingProfileId(id.billingProfileId, samRequestContext)
+      mrgCoordinates <- mrgOpt match {
+        case Some(mrg) => IO.pure(mrg.managedResourceGroupCoordinates)
+        case None =>
+          IO.raiseError(
+            new WorkbenchExceptionWithErrorReport(
+              ErrorReport(StatusCodes.NotFound, s"Managed Resource Group with Billing Profile ID [${id.billingProfileId}] does not exist")
+            )
+          )
+      }
+      _ <- validateManagedResourceGroup(mrgCoordinates, samRequestContext)
```

Comment:
Make sure that this actually prevents resources in azure from being created, note concurrency in for yields